kpi/models/kpi.py

```diff
@@ -0,0 +1,205 @@
+    kpi_type = fields.Selection((
+        ('python', 'Python'),
+        ('local', 'SQL - Local DB'),
+        ('external', 'SQL - External DB')
+    ), 'KPI Computation Type')
+
+    dbsource_id = fields.Many2one(
+        'base.external.dbsource',
+        'External DB Source',
+    )
+    kpi_code = fields.Text(
+        'KPI Code',
+        help=("SQL code must return the result as 'value' "
+              "(i.e. 'SELECT 5 AS value')."),
+    )
+    history_ids = fields.One2many(
+        'kpi.history',
+        'kpi_id',
+        'History',
+    )
+    active = fields.Boolean(
+        'Active',
+        help=("Only active KPIs will be updated by the scheduler based on"
+              " the periodicity configuration."), default=True
+    )
+    company_id = fields.Many2one(
+        'res.company', 'Company',
+        default=lambda self: self.env.user.company_id.id)
+
+    @api.multi
+    def _compute_display_last_kpi_value(self):
+        history_obj = self.env['kpi.history']
+        for obj in self:
+            history_ids = history_obj.search([("kpi_id", "=", obj.id)])
+            if history_ids:
+                his = obj.history_ids[0]
+                obj.value = his.value
+                obj.color = his.color
+                obj.last_execution = his.date
+            else:
+                obj.value = 0
+                obj.color = '#FFFFFF'
+                obj.last_execution = False
+
+    @api.multi
+    def _get_kpi_value(self):
+        self.ensure_one()
+        kpi_value = 0
+        if self.kpi_code:
+            if self.kpi_type == 'local' and is_sql_or_ddl_statement(
+                    self.kpi_code):
+                self.env.cr.execute(self.kpi_code)
+                dic = self.env.cr.dictfetchall()
+                if is_one_value(dic):
+                    kpi_value = dic[0]['value']
+            elif (self.kpi_type == 'external' and self.dbsource_id.id and
+                  is_sql_or_ddl_statement(self.kpi_code)):
+                dbsrc_obj = self.dbsource_id
+                res = dbsrc_obj.execute(self.kpi_code)
+                if is_one_value(res):
+                    kpi_value = res[0]['value']
+            elif self.kpi_type == 'python':
+                kpi_value = safe_eval(self.kpi_code, {'self': self})
+        if isinstance(kpi_value, dict):
+            res = kpi_value
+        else:
+            threshold_obj = self.threshold_id
+            res = {
+                'value': kpi_value,
+                'color': threshold_obj.get_color(kpi_value),
+            }
+        res.update({'kpi_id': self.id})
+        return res
+
+    @api.multi
+    def compute_kpi_value(self):
+        for obj in self:
+            history_vals = obj._get_kpi_value()
+            history_obj = self.env['kpi.history']
+            history_obj.create(history_vals)
+        return True
+
+    @api.multi
+    def update_next_execution_date(self):
+        for obj in self:
+            if obj.periodicity_uom == 'hour':
```

Comment:
It makes sense to add the option of minutes periodicity?

Reply:
Now added!